### ISSUE TITLE
Unix permissions are messed up after upgrading to 2.6.0+

### ISSUE THREAD

**Issue by suushiemaniac** [srikanth-lingala/zip4j#231]
OS: Manjaro 20.1 / Linux 5.7.14
Java: 11.0.8
Zip4j: 2.6.0 / 2.6.1 (both produce the issue)

After upgrading my zip4j version, I cannot open files from the ZIP any more because my OS claims that "permission denied". Upon further investigation, the files all have permission `?---------` instead of the usual `-rw----`

I am not very knowledgeable about ZIP headers, but my wild guess is [this commit](https://github.com/srikanth-lingala/zip4j/commit/76db682d631052bb1e842a6570fc00cb00c28264). I have attached a debugger and the `setVersionMadeBy` value changed from the constant `20` to `819` (as the result of `determineVersionMadeBy` invocation).

Note that my archive viewer also shows different values for the `OS` and `Version` columns (OS is "Betriebssystem" in German). Columns translate as follows:
Original size | Compressed size | Saved space | Date | Time | Permissions | Method | Operating System | Version

## **before** with Zip4j 2.5.2
![2020-08-21-183600_750x86_scrot](https://user-images.githubusercontent.com/6136469/90914155-b01a1b80-e3dd-11ea-8c1b-857f16eae11c.png)
## **after** with Zip4j 2.6.0 / 2.6.1
![2020-08-21-183759_754x85_scrot](https://user-images.githubusercontent.com/6136469/90914157-b0b2b200-e3dd-11ea-8e4a-6c70a9c76240.png)

I'd expect that I can continue to open files within the ZIP as usual. At first I was considering that it's the fault of my archive viewer, but please note that a user of my app (which in turn generates ZIP files with your library) reported the same problem under a different operating system: https://github.com/thewca/tnoodle/issues/600

**Comment by srikanth-lingala** [comment]
If I understand your issue right: you/your users have created the zip file with zip4j, and are trying to view/extract with a third party tool? If my understanding is right, what is the behaviour when you extract the zip file with zip4j? Do you see the proper file permissions being restored? 

Message below is with the presumption that my understanding of the issue as mentioned above is right.

I am unable to reproduce this on a mac. I extracted Linux261.zip (from your last message), both with zip4j and mac default tool, and in each occasion, I see the right permissions being applied. 

![image](https://user-images.githubusercontent.com/3450302/91032706-a7565f00-e602-11ea-8684-79de5eeba1a1.png)

Unfortunately, I do not have access to a linux machine, so we might have to do a bit of coordination to investigate this issue. First task now is to see if the issue is with the archive viewer or with zip4j. I have my doubts that the issue is with zip4j. I might be wrong, but I think that the issue might as well be with the archive viewer/extractor that you are using. I made a zip file with Keka (a zip utility on mac). Please find that attached. Can you please try to extract that with your archiver and see if you see the permissions in this case?

[ZipLinux261_Keka.zip](https://github.com/srikanth-lingala/zip4j/files/5117164/ZipLinux261_Keka.zip)

**Comment by suushiemaniac** [comment]
Thanks for digging into this issue! Your understanding of the issue is correct. I have yet to attempt re-extracting the archive using zip4j. Will report back as soon as I get the chance!

I can eliminate the concern that the problem is only the fault of my third-party ZIP viewer... Using the `unzip` CLI command (which is the most basic, GUI-agnostic approach to extracting ZIPs on Linux) produces the same erroneous file permissions after extraction.

As you said you do not have access to a Linux machine, I went ahead and dumped some `zipinfo` outputs. I will post them below, with the according filenames attached.

```
$ zipinfo ZipLinux252.zip
Archive:  ZipLinux252.zip
Zip file size: 51497 bytes, number of entries: 8
-rw----     2.0 fat    18392 bl defN 20-Aug-22 11:51 Printing/Scramble Sets/3x3x3 Round 1.pdf
-rw----     2.0 fat    18611 bl defN 20-Aug-22 11:51 Printing/ZipLinux252 - All Scrambles.pdf
-rw----     2.0 fat      365 bl defN 20-Aug-22 11:51 Interchange/txt/3x3x3 Round 1.txt
-rw----     2.0 fat      939 bl defN 20-Aug-22 11:51 Interchange/ZipLinux252.json
-rw----     2.0 fat      961 bl defN 20-Aug-22 11:51 Interchange/ZipLinux252.jsonp
-rw----     2.0 fat      537 bl defN 20-Aug-22 11:51 Interchange/ZipLinux252.html
-rw----     2.0 fat    16676 bl defN 20-Aug-22 11:51 ZipLinux252 - Computer Display PDFs.zip
-rw----     2.0 fat      335 bl defN 20-Aug-22 11:51 ZipLinux252 - Computer Display PDF Passcodes - SECRET.txt
8 files, 56816 bytes uncompressed, 50151 bytes compressed:  11.7%
```

```
$ zipinfo ZipLinux261.zip
Archive:  ZipLinux261.zip
Zip file size: 51784 bytes, number of entries: 8
?---------  5.1 unx    18382 bl defN 20-Aug-22 11:56 Printing/Scramble Sets/3x3x3 Round 1.pdf
?---------  5.1 unx    18601 bl defN 20-Aug-22 11:56 Printing/ZipLinux261 - All Scrambles.pdf
?---------  5.1 unx      370 bl defN 20-Aug-22 11:56 Interchange/txt/3x3x3 Round 1.txt
?---------  5.1 unx      944 bl defN 20-Aug-22 11:56 Interchange/ZipLinux261.json
?---------  5.1 unx      966 bl defN 20-Aug-22 11:56 Interchange/ZipLinux261.jsonp
?---------  5.1 unx      537 bl defN 20-Aug-22 11:56 Interchange/ZipLinux261.html
?---------  5.1 unx    16665 bl defN 20-Aug-22 11:56 ZipLinux261 - Computer Display PDFs.zip
?---------  5.1 unx      335 bl defN 20-Aug-22 11:56 ZipLinux261 - Computer Display PDF Passcodes - SECRET.txt
8 files, 56800 bytes uncompressed, 50438 bytes compressed:  11.2%
```

```
$ zipinfo ZipWindows252.zip
Archive:  ZipWindows252.zip
Zip file size: 51576 bytes, number of entries: 8
-rw----     2.0 fat    18360 bl defN 20-Aug-22 12:25 Printing/Scramble Sets/3x3x3 Round 1.pdf
-rw----     2.0 fat    18579 bl defN 20-Aug-22 12:25 Printing/ZipWindows252 - All Scrambles.pdf
-rw----     2.0 fat      355 bl defN 20-Aug-22 12:25 Interchange/txt/3x3x3 Round 1.txt
-rw----     2.0 fat      937 bl defN 20-Aug-22 12:25 Interchange/ZipWindows252.json
-rw----     2.0 fat      959 bl defN 20-Aug-22 12:25 Interchange/ZipWindows252.jsonp
-rw----     2.0 fat      539 bl defN 20-Aug-22 12:25 Interchange/ZipWindows252.html
-rw----     2.0 fat    16635 bl defN 20-Aug-22 12:25 ZipWindows252 - Computer Display PDFs.zip
-rw----     2.0 fat      337 bl defN 20-Aug-22 12:25 ZipWindows252 - Computer Display PDF Passcodes - SECRET.txt
8 files, 56701 bytes uncompressed, 50206 bytes compressed:  11.5%
```

```
$ zipinfo ZipWindows261.zip
Archive:  ZipWindows261.zip
Zip file size: 51780 bytes, number of entries: 8
-rw----     5.1 fat    18361 bl defN 20-Aug-22 12:26 Printing/Scramble Sets/3x3x3 Round 1.pdf
-rw----     5.1 fat    18580 bl defN 20-Aug-22 12:26 Printing/ZipWindows261 - All Scrambles.pdf
-rw----     5.1 fat      366 bl defN 20-Aug-22 12:26 Interchange/txt/3x3x3 Round 1.txt
-rw----     5.1 fat      947 bl defN 20-Aug-22 12:26 Interchange/ZipWindows261.json
-rw----     5.1 fat      969 bl defN 20-Aug-22 12:26 Interchange/ZipWindows261.jsonp
-rw----     5.1 fat      539 bl defN 20-Aug-22 12:26 Interchange/ZipWindows261.html
-rw----     5.1 fat    16648 bl defN 20-Aug-22 12:26 ZipWindows261 - Computer Display PDFs.zip
-rw----     5.1 fat      337 bl defN 20-Aug-22 12:26 ZipWindows261 - Computer Display PDF Passcodes - SECRET.txt
8 files, 56747 bytes uncompressed, 50410 bytes compressed:  11.2%
```

Observations: For either OS, the ZIP specification version bumps from `2.0` to `5.1`. Version `2.5.2` produces `fat` file entries even when generated on a Linux machine.

Here's that same ZIP info for your manually created Keka archive:

```
$ zipinfo ZipLinux261_Keka.zip
Archive:  ZipLinux261_Keka.zip
Zip file size: 52960 bytes, number of entries: 13
drwxr-xr-x  6.3 unx        0 bx stor 20-Aug-24 11:50 ZipLinux261/
drwxr-xr-x  6.3 unx        0 bx stor 20-Aug-24 11:50 ZipLinux261/Interchange/
-rw-rw-r--  6.3 unx      537 bx defN 20-Aug-22 11:56 ZipLinux261/Interchange/ZipLinux261.html
-rw-rw-r--  6.3 unx      944 bx defN 20-Aug-22 11:56 ZipLinux261/Interchange/ZipLinux261.json
-rw-rw-r--  6.3 unx      966 bx defN 20-Aug-22 11:56 ZipLinux261/Interchange/ZipLinux261.jsonp
drwxr-xr-x  6.3 unx        0 bx stor 20-Aug-24 11:50 ZipLinux261/Interchange/txt/
-rw-rw-r--  6.3 unx      370 bx defN 20-Aug-22 11:56 ZipLinux261/Interchange/txt/3x3x3 Round 1.txt
drwxr-xr-x  6.3 unx        0 bx stor 20-Aug-24 11:50 ZipLinux261/Printing/
drwxr-xr-x  6.3 unx        0 bx stor 20-Aug-24 11:50 ZipLinux261/Printing/Scramble Sets/
-rw-rw-r--  6.3 unx    18382 bx defN 20-Aug-22 11:56 ZipLinux261/Printing/Scramble Sets/3x3x3 Round 1.pdf
-rw-rw-r--  6.3 unx    18601 bx defN 20-Aug-22 11:56 ZipLinux261/Printing/ZipLinux261 - All Scrambles.pdf
-rw-rw-r--  6.3 unx      335 bx defN 20-Aug-22 11:56 ZipLinux261/ZipLinux261 - Computer Display PDF Passcodes - SECRET.txt
-rw-rw-r--  6.3 unx    16665 bx stor 20-Aug-22 11:56 ZipLinux261/ZipLinux261 - Computer Display PDFs.zip
13 files, 56800 bytes uncompressed, 50462 bytes compressed:  11.2%
```

As you can see, the permissions are there but then again the archive is also based on `bx` folder hierarchies instead of `bl` nodes (be aware that I only have a very rudimentary understanding of what this means, but I'm digging through the ZIP specification documents to get a better understanding :grin: )

**Comment by srikanth-lingala** [comment]
Thanks for the hint with zipinfo command. Although I am able to extract the zip file with right permissions with the default mac zip tool, I am able to see the same output on my mac with that zipinfo command as well (with question marks in permissions with 261). This will at least save some time with back and forth communication.

**Comment by suushiemaniac** [comment]
I did some more digging, and I found the following remark on the man-page for `unzip` (cf. `man unzip`):
> Dates, times and permissions of stored directories are not restored except under Unix. (On Windows NT and successors, timestamps are now restored.)

Note the part that says "except under Unix". If you use `zipinfo` with an additional `-v` flag it will give you (much!) more detailed information. I tried to leverage that information by extracting the Linux261 file, then **manually** `chmod`ding all non-directory file contents to `0644` and then packing it with the CLI `zip` utility. The result comparisons are as follows:

ZipLinux261
```
Central directory entry #1:
---------------------------

  Printing/Scramble Sets/3x3x3 Round 1.pdf

  offset of local header from start of archive:   0
                                                  (0000000000000000h) bytes
  file system or operating system of origin:      Unix
  version of encoding software:                   5.1
  minimum file system compatibility required:     MS-DOS, OS/2 or NT FAT
  minimum software version required to extract:   2.0
  compression method:                             deflated
  compression sub-type (deflation):               normal
  file security status:                           not encrypted
  extended local header:                          yes
  file last modified on (DOS date/time):          2020 Aug 22 11:56:42
  32-bit CRC value (hex):                         d9492396
  compressed size:                                16004 bytes
  uncompressed size:                              18382 bytes
  length of filename:                             40 characters
  length of extra field:                          0 bytes
  length of file comment:                         0 characters
  disk number on which file begins:               disk 1
  apparent file type:                             binary
  Unix file attributes (000000 octal):            ?---------
  MS-DOS file attributes (00 hex):                none

  There is no file comment.
```

ZipLinuxManual (the one I created by extracting, chmodding, zipping again)
```
Central directory entry #8:
---------------------------

  Printing/Scramble Sets/3x3x3 Round 1.pdf

  offset of local header from start of archive:   1827
                                                  (0000000000000723h) bytes
  file system or operating system of origin:      Unix
  version of encoding software:                   3.0
  minimum file system compatibility required:     MS-DOS, OS/2 or NT FAT
  minimum software version required to extract:   2.0
  compression method:                             deflated
  compression sub-type (deflation):               normal
  file security status:                           not encrypted
  extended local header:                          no
  file last modified on (DOS date/time):          2020 Aug 22 11:56:42
  file last modified on (UT extra field modtime): 2020 Aug 22 11:56:42 local
  file last modified on (UT extra field modtime): 2020 Aug 22 09:56:42 UTC
  32-bit CRC value (hex):                         d9492396
  compressed size:                                16003 bytes
  uncompressed size:                              18382 bytes
  length of filename:                             40 characters
  length of extra field:                          24 bytes
  length of file comment:                         0 characters
  disk number on which file begins:               disk 1
  apparent file type:                             binary
  Unix file attributes (100644 octal):            -rw-r--r--
  MS-DOS file attributes (00 hex):                none

  The central-directory extra field contains:
  - A subfield with ID 0x5455 (universal time) and 5 data bytes.
    The local extra field has UTC/GMT modification/access times.
  - A subfield with ID 0x7875 (Unix UID/GID (any size)) and 11 data bytes:
    01 04 e8 03 00 00 04 e9 03 00 00.

  There is no file comment.
```

Here's my guess at what happens: Until `2.5.1` you were writing a header that implied `MS/DOS` as per section `4.4.2.2` of [the ZIP specification](https://pkware.cachefly.net/webdocs/casestudies/APPNOTE.TXT). The Linux archivers then by default do not even _try_ to restore file permissions because they know it's not going to work (coming from a fat-esque FS) and assume `-rw----` as kind of a "sane default".
Now with the `2.6.0` update you started writing `unix` headers, so the unzip utility goes like "Hey, that's something that I can restore permissions from!". But as the _actual_ permissions are never stored (cf. my `zipinfo -v` extract from above) it fails to restore them upon extraction.

Suggested solution: Either (a) investigate how to write Unix file attributes or (b) kinda short-circuit the issue by allowing the user to force Windows mode.

Hope my explanation makes sense!

**Comment by suushiemaniac** [comment]
In the linked document about the ZIP specification, it says under `4.4.15`:

> 4.4.15 external file attributes: (4 bytes)
>
>       The mapping of the external attributes is
>       host-system dependent (see 'version made by').  For
>       MS-DOS, the low order byte is the MS-DOS directory
>       attribute byte.  If input came from standard input, this
>       field is set to zero.

Maybe that's where to write the file permissions? 4 bytes would *theoretically* be enough for the 6*3 octal groups of Unix file permissions

**Comment by suushiemaniac** [comment]
Scrap that previous comment. It turns out that the `zip` and `unzip` CLI commands were developed by a company called `Info-ZIP` which defined their own extra fields for file permissions.

Their documentation is at ftp://ftp.info-zip.org/pub/infozip/doc/ and the latest version extensively describes how to encode UNIX file permissions as extra field in the file headers for every entry. Quote:

> -Info-ZIP Unix Extra Field (type 2):
>          ==================================
>
>          The following is the layout of the new Info-ZIP extra block for
>          Unix.  (Last Revision 19960922)
>
>          Local-header version:
>
>          Value         Size        Description
>          -----         ----        -----------
>  (Unix2) 0x7855        Short       tag for this extra block type ("Ux")
>          TSize         Short       total data size for this block (4)
>          UID           Short       Unix user ID
>          GID           Short       Unix group ID
>
>          Central-header version:
>
>          Value         Size        Description
>          -----         ----        -----------
>  (Unix2) 0x7855        Short       tag for this extra block type ("Ux")
>          TSize         Short       total data size for this block (0)
>
>          The data size of the central-header version is zero; it is used
>          solely as a flag that UID/GID info is present in the local-header
>          extra field.  If additional fields are ever added to the local
>          version, the central version may be extended to indicate this.
>
>          Note that Unix UIDs and GIDs are usually specific to a particular
>          machine, and they generally require root access to restore.
>
>
>         -ASi Unix Extra Field:
>          ====================
>
>          The following is the layout of the ASi extra block for Unix.  The
>          local-header and central-header versions are identical.
>          (Last Revision 19960916)
>
>          Value         Size        Description
>          -----         ----        -----------
>  (Unix3) 0x756e        Short       tag for this extra block type ("nu")
>          TSize         Short       total data size for this block
>          CRC           Long        CRC-32 of the remaining data
>          Mode          Short       file permissions
>          SizDev        Long        symlink'd size OR major/minor dev num
>          UID           Short       user ID
>          GID           Short       group ID
>          (var.)        variable    symbolic link filename
>
>          Mode is the standard Unix st_mode field from struct stat, containing
>          user/group/other permissions, setuid/setgid and symlink info, etc.
>
>          If Mode indicates that this file is a symbolic link, SizDev is the
>          size of the file to which the link points.  Otherwise, if the file
>          is a device, SizDev contains the standard Unix st_rdev field from
>          struct stat (includes the major and minor numbers of the device).
>          SizDev is undefined in other cases.
>
>          If Mode indicates that the file is a symbolic link, the final field
>          will be the name of the file to which the link points.  The file-
>          name length can be inferred from TSize.
>
>          [Note that TSize may incorrectly refer to the data size not counting
>           the CRC; i.e., it may be four bytes too small.]

I never thought I'd say this in my entire life, but perhaps here falling back to the Windows/DOS approach might be the easiest solution :sweat_smile:

**Comment by srikanth-lingala** [comment]
I think I figured out the issue. The big clue came to me in a screenshot you posted of the output of `zipinfo` command. The entries in the zip created by zip4j by using ZipFile api should have "b-" in one of the columns of the output, instead it was "bl". Having "bl" means that the entry has an extra data record associated with it, which, in case of zip file created with zip4j means that the file was added by ZipOutputStream and not the ZipFile api. I looked up the code in your repo, and [it seems][2] that you are using ZipOutputStream to create the zip file.

And in fact, I was able to reproduce this issue only when using ZipOutputStream (ZipFile api works fine). The issue is that when using ZipOutputStream, since it cannot determine the permission of the stream being added, zip4j does not set any file permissions (technically speaking it sets the value to 0). zip4j version <= 2.5.2 did not set the right versionMadeBy and versionNeededToExtract, and I presume (as you also did in one of your post above), the tool defaults to some basic permissions in that case. I will add a fix to write some default permissions in case of adding files via stream.

To answer the points you made in the posts above: Yes, InfoZip has added some extra headers for unix permissions, but zip specification also has a way to store these permissions in the header. This can be calculated by some complicated logic. If you are interested, you can find the logic in [FileUtils.getPosixFileAttributes()][1].

[1]: https://github.com/srikanth-lingala/zip4j/blob/master/src/main/java/net/lingala/zip4j/util/FileUtils.java#L465
[2]: https://github.com/thewca/tnoodle/blob/master/webscrambles/src/main/kotlin/org/worldcubeassociation/tnoodle/server/webscrambles/zip/model/ZipArchive.kt#L25

**Comment by suushiemaniac** [comment]
Thank you for your thorough explanation and for promptly thinking of a fix! Reading all these interesting specifications about the ZIP format was so fascinating, maybe I went a bit overboard with my comments. Sorry! 😅

I also understand the fundamental issue that the ZipOutputStream cannot possibly know the file permissions of the content that is written. My problem is that there's absolutely no way for me to write to a Java `File` (through the ZipFile implementation) because that has to actually exist somewhere on the disk, outside of the program itself.

In my specific use-case I only require the bytes in-memory and I cannot even afford to write stuff to a temporary file, read its bytes and then remove the temp file again. Therefore I'm relieved to hear that you're adding a fix :)

Thanks and keep up the great work!

**Comment by srikanth-lingala** [comment]
> maybe I went a bit overboard with my comments.

Absolutely not. I really appreciate that you actively contributed to the investigation. In fact it is your parallel investigation and effort that really helped me to find out the root cause of the issue. As I wrote in my post above, you have talked in one of your post about "bl" and "bx" and attached a screenshot. That clue was all I needed to go in the right path and figure out the issue. So, thanks a lot for your support.

I totally understand that your use case requires you to use ZipOutputStream, and it is perfectly fine using ZipOutputStream. ZipOutputStream is there for exactly such purposes when you don't want to (and have to) deal with files. It does not make sense to write content to a temporary file just to add it to a zip file and then later delete the temporary file. A lot of overhead and unwanted processing.

I will include the fix in the next release. I only have one more issue to fix, and then the next release is ready (unless nothing new comes up in the next couple of days).

**Comment by srikanth-lingala** [comment]
Issue fixed in v2.6.2 released today.

**Comment by hugomprado** [comment]
Excellent bug tracking and fix resolution!
I'm experiencing the same issue though. I read all the investigation and sympthoms and I was able to reproduce and correct the error on a unix server. However, we have a HP-UX box which is running same code and only here the code is failling.
Srikanth, if you have some time, could you take a look? Maybe the line: OPERATING_SYSTEM_NAME.contains("nux") should be extended to embrace old unix falvors also.
Thanks and regards.
